src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,700 @@
+                          0, m_dst_image_ctx->layout.object_size, image_extents);
+
+  size_t total = 0;
+  for (auto &e : image_extents) {
+    std::map<object_t, std::vector<ObjectExtent>> src_object_extents;
+    Striper::file_to_extents(m_cct, m_src_image_ctx->format_string,
+                             &m_src_image_ctx->layout, e.first, e.second, 0,
+                             src_object_extents);
+    for (auto &p : src_object_extents) {
+      for (auto &s : p.second) {
+        m_src_objects.insert(s.objectno);
+        m_src_object_extents.push_back({s.objectno, s.offset, s.length});
+        total += s.length;
+      }
+    }
+  }
+
+  assert(total == m_dst_image_ctx->layout.object_size);
+
+  ldout(m_cct, 20) << m_src_object_extents.size() << " src extents" << dendl;
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::compute_read_ops() {
+  librados::snap_t src_copy_point_snap_id = m_snap_map.rbegin()->first;
+  uint64_t prev_end_size = 0;
+  bool prev_exists = false;
+  librados::snap_t start_src_snap_id = 0;
+
+  m_read_ops = {};
+  m_read_snaps = {};
+  m_zero_interval = {};
+  m_end = {};
+  m_state = {};
+
+  for (auto &pair : m_snap_map) {
+    assert(!pair.second.empty());
+    librados::snap_t end_src_snap_id = pair.first;
+    librados::snap_t end_dst_snap_id = pair.second.front();
+
+    interval_set<uint64_t> diff;
+    uint64_t end_size;
+    bool exists;
+    librados::snap_t clone_end_snap_id;
+    calc_snap_set_diff(m_cct, m_snap_set, start_src_snap_id,
+                       end_src_snap_id, &diff, &end_size, &exists,
+                       &clone_end_snap_id);
+
+    ldout(m_cct, 20) << "start_src_snap_id=" << start_src_snap_id << ", "
+                     << "end_src_snap_id=" << end_src_snap_id << ", "
+                     << "clone_end_snap_id=" << clone_end_snap_id << ", "
+                     << "end_dst_snap_id=" << end_dst_snap_id << ", "
+                     << "diff=" << diff << ", "
+                     << "end_size=" << end_size << ", "
+                     << "exists=" << exists << dendl;
+
+    if (exists) {
+      // clip diff to size of object (in case it was truncated)
+      if (end_size < prev_end_size) {
+        interval_set<uint64_t> trunc;
+        trunc.insert(end_size, prev_end_size);
+        trunc.intersection_of(diff);
+        diff.subtract(trunc);
+        ldout(m_cct, 20) << "clearing truncate diff: " << trunc << dendl;
+      }
+
+      // reads should be issued against the newest (existing) snapshot within
+      // the associated snapshot object clone. writes should be issued
+      // against the oldest snapshot in the snap_map.
+      assert(clone_end_snap_id >= end_src_snap_id);
+      if (clone_end_snap_id > src_copy_point_snap_id) {
+        // do not read past the copy point snapshot
+        clone_end_snap_id = src_copy_point_snap_id;
+      }
+
+      m_state[end_src_snap_id] = OBJECT_EXISTS;
```

Comment:
Why do you need to track the "source object" state? The destination object state should be computable from the fact that (1) the object already exists in dest from a previous snapshot zero write ops, or (2) you had one or more write ops.